src/app/core/uri-retriever.service.ts

```diff
@@ -0,0 +1,245 @@
+import { Injectable } from '@angular/core';
+import { HttpClient, HttpErrorResponse } from '@angular/common/http';
+import { Observable, throwError, Subject } from 'rxjs';
+import { LearningObject } from '@entity';
+import { USER_ROUTES, PUBLIC_LEARNING_OBJECT_ROUTES } from '@env/route';
+import { LearningOutcome } from '@entity';
+import { catchError, retry, map, tap, filter, take, takeUntil, finalize } from 'rxjs/operators';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class UriRetrieverService {
+
+  constructor(private http: HttpClient) { }
+
+  /**
+   * This function will return the metadata and the requested resources for a learning object as a promise of a Learning Object.
+   * @params params{ author: string, name: string, id:string } the values needed to retrieve the metadata for a learning object
+   * @params resources the resources (i.e. children, parents, outcomes, etc) that need to be loaded with the metadata
+   *
+   *
+   */
+  async getLearningObject(
+    params: { author?: string, name?: string, id?: string },
+    resources?: string[]
+  ): Promise<LearningObject> {
+    let properties: string[];
+    if (resources) {
+      properties = resources;
+    } else {
+      properties = [];
+    }
+
+    const learningObject = {};
+    const request = this.getLearningObjectResources(params, properties);
+
+    return new Promise((resolve) => {
+      request.pipe(
+        finalize(() => resolve(learningObject as LearningObject))
+      ).subscribe(val => {
+        if (val.requestKey) {
+          learningObject[val.requestKey] = val.value;
+        } else {
+          const object = (val as LearningObject).toPlainObject();
+
+          // tslint:disable-next-line: forin
+          for (const key in object) {
+            learningObject[key] = object[key];
+          }
+        }
+      });
+    });
+  }
+
+  /**
+   * This function will return a subject that includes the requested learning objects metadata and
+   * all of the requested properties.
+   * @param params the author, name, or id needed to retrieve the metadata for a learning object
+   * @param properties the properties (i.e. children, parents, etc) that have been requested
+   */
+  getLearningObjectResources(
+    params: { author?: string, name?: string, id?: string },
+    properties?: string[]
+  ) {
+    let route;
+
+    if (params.id) {
+      route = USER_ROUTES.GET_LEARNING_OBJECT(params.id);
+    } else {
+      route = PUBLIC_LEARNING_OBJECT_ROUTES.GET_PUBLIC_LEARNING_OBJECT(params.author, params.name);
+    }
```

Comment:
This function can be broken up to smaller helper functions. Doing so will improve readability. It will also making testing easier. For example, if you move lines 67 - 71 into a function, then you can write to tests to ensure that the correct route structure is returned.